Cuprite support

myabc opened the pull request:
## Summary

Adds [Cuprite](https://github.com/rubycdp/cuprite) (Ferrum/CDP-based headless Chrome) as a supported Capybara driver alongside Selenium and rack_test.

The accessibility methods (`accessible_name`, `accessible_description`, `role`) and the `:role` selector previously only worked under Selenium and rack_test. This wires up a Cuprite path for each, resolving computed values from Chrome's accessibility tree via CDP (`Accessibility.getPartialAXTree`), and fixes a couple of driver-behaviour differences that affected the disclosure and rich-text selectors.

All existing specs run green on Cuprite — there are **no Cuprite-skipped specs**.

## What's included

**Driver wiring**
- New `Cuprite::AccessibilityComputedValue` resolves a node's computed name/role/description from the CDP accessibility tree.
- `accessible_name` / `role` read those computed values; `accessible_description` reuses the existing accname-1.2 description algorithm (it only depends on the generic Capybara node API, so it is driver-agnostic) — this keeps description results consistent with Selenium rather than returning Chrome's raw value.
- Cuprite extensions are mixed in only when Cuprite is loaded, so it is **not** a runtime dependency of the gem.

**Cross-driver behaviour fixes**
- **Disclosure** — `details[:open]` is a boolean attribute whose raw value differs per driver (Selenium `"true"`/`nil`, Cuprite `true`/`false`). The summary toggle now normalises it before comparing, so re-toggling an already-open disclosure no longer misfires under Cuprite.
- **Rich text** — Cuprite/Ferrum cannot reliably trigger a select-all chord (mod+a) inside a `contenteditable`, so `fill_in_rich_text`'s clear was a no-op and text was appended instead of replaced. Under Cuprite it now moves the caret to the end and backspaces through the content; Selenium keeps the faster chord.
- Minor robustness: the CDP resolver guards a missing accessibility tree / CDP error and degrades to an empty value, mirroring the Selenium rescue paths; `accessible_name` collapses whitespace to match the Selenium path.

**Tooling / docs**
- CI now runs the suite against Cuprite (`DRIVER=cuprite_chrome_headless`) in addition to Chromium, Firefox, and rack_test.
- README documents that Cuprite is optional and **must be required before `capybara_accessible_selectors`** for the extensions to be applied.

## Load-order requirement

Because Cuprite is optional, its extensions are only installed if `Capybara::Cuprite::Node` is already defined at require time:

```ruby
require "capybara/cuprite"
require "capybara_accessible_selectors"
```

This is documented in the README. A lazy/registration-based hook to remove the ordering constraint was considered but intentionally left out of scope.

## Test plan

- [x] `bundle exec rspec` (default Selenium Chromium driver) — 0 failures
- [x] `DRIVER=selenium_headless bundle exec rspec` (Firefox) — 0 failures
- [x] `DRIVER=cuprite_chrome_headless bundle exec rspec` — 0 failures, **0 skips**
- [x] `DRIVER=rack_test bundle exec rspec` — 0 failures
- [x] `bundle exec rubocop`


## Upstream Ferrum API (follow-up)

1. Ferrum Cuprite Support 
   The Cuprite path currently reaches into raw CDP (`Accessibility.getPartialAXTree`) because Ferrum exposes no public accessibility API. [rubycdp/ferrum#595](https://github.com/rubycdp/ferrum/pull/595) adds a first-class `page.accessibility` domain plus a `node.axnode` convenience returning an `AXNode` value object (`name`/`role`/`description`).

   Validated locally against that branch: `Cuprite::AccessibilityComputedValue` collapses to `node.axnode&.public_send(name)`, dropping the raw-CDP coupling, ignored-node filtering, and hash-digging. Full suite stays green (1813 examples, 0 failures).

   Once #595 ships in a Ferrum release, this PR's resolver can be simplified and the dependency constraint bumped as a follow-up.
2. Ferrum key handling (select all) fix: https://github.com/rubycdp/ferrum/pull/600

myabc commented on lib/capybara_accessible_selectors/node.rb:
```diff
@@ -3,11 +3,13 @@
 
   ::Capybara::Driver::Node.include DriverNodeExtensions
   ::Capybara::Node::Element.include NodeElementExtensions
+  ::Capybara::Cuprite::Node.include CupriteNodeExtensions
```
As implemented so far, this would make Cuprite a dependency. This is probably not what we want.

seanpdoyle replied:
Would adding a `defined?` guard clause suffice?


```suggestion
  ::Capybara::Cuprite::Node.include(CupriteNodeExtensions) if defined?(::Capybara::Cuprite::Node)
```

myabc commented on spec/spec_helper.rb:
```diff
@@ -78,6 +79,9 @@ def squish
   Capybara::Selenium::Driver.new(app, browser: :chrome, options:)
 end
+Capybara.register_driver(:cuprite_chrome_headless) do |app|
+  Capybara::Cuprite::Driver.new(app, browser_name: :chrome)
```
Options listed here: https://github.com/rubycdp/ferrum#customization

danieldevlewis commented on lib/capybara_accessible_selectors/node/accessible_description.rb:
```diff
@@ -16,6 +17,12 @@ def accessible_description
 
+  module CupriteNodeExtensions
+    def accessible_description
+      Cuprite::AccessibilityComputedValue.resolve(native, "description") || ""
```
[The specification](https://www.w3.org/TR/accname-1.2/#mapping_additional_nd_description) is rather loose and each browser has implemented it slightly differently.

There wasn't an `accessible_description` method for selenium so it is just my interpretation.

So I'd say, as long as it is returning accessible descriptions, passing these edge cases isn't important.

myabc commented:
@seanpdoyle thanks for your feedback and apologies for dropping the ball! I'd still love to see Cuprite support. I'll see if I can address the various comments in the next week or so.

myabc committed:
Add Cuprite as dev/test dependency

myabc committed:
Implement Cuprite driver support

Add Cuprite (Chrome DevTools Protocol) as an alternative browser driver.
Uses Chrome's Accessibility.getPartialAXTree CDP command to resolve
accessible name, description, and role values.

- Add CupriteNodeExtensions with defined? guard for optional dependency
- Add AccessibilityComputedValue resolver for CDP accessibility tree
- Handle Chrome's non-standard PascalCase role identifiers
- Guard Cuprite require/registration in spec_helper
- Use RUBY_PLATFORM for platform detection (no hard Selenium dependency)
- Use evaluate_script for outerHTML access across drivers
- Add send_keys nil guards for rich text

Fixes #156

myabc committed:
Skip specs incompatible with Cuprite driver

Annotate specs with skip_driver: :cuprite_chrome where Cuprite's
send_keys handling of modifier keys differs from Selenium, or where
the <details> element interaction behaves differently.

- 2 disclosure toggle/select specs (details element interaction)
- 8 rich text fill/replace/within specs (modifier key handling)

myabc commented on lib/capybara_accessible_selectors/selectors/rich_text.rb:
```diff
@@ -42,7 +42,7 @@ def fill_in_rich_text(locator, with:, clear: true, **find_options)
         # So we can't just use backspace
         input.send_keys [command_modifier, "a"], :backspace if input.text != "" && clear
-        input.send_keys with
+        input.send_keys with if with
```
Note to self: would be good to fix upstream!

myabc commented on lib/capybara_accessible_selectors/node.rb:
```diff
@@ -3,11 +3,13 @@
   module RackTestNodeExtensions; end
 
   ::Capybara::Driver::Node.include DriverNodeExtensions
   ::Capybara::Node::Element.include NodeElementExtensions
+  ::Capybara::Cuprite::Node.include(CupriteNodeExtensions) if defined?(::Capybara::Cuprite::Node)
   ::Capybara::Selenium::Node.include SeleniumNodeExtensions
   ::Capybara::RackTest::Node.include RackTestNodeExtensions
   ::Capybara::Node::Simple.include RackTestNodeExtensions
```
This suggestion is much a more elaborate approach to loading drivers - which I won't try to address here.

seanpdoyle replied:
While the alternative seems to be much, much more complicated, the concern is correct.

I was unable to pass my `:role`-reliant tests without making this change (note the change in order):

```diff
   gem "capybara"
-  gem "capybara_accessible_selectors", github: "citizensadvice/capybara_accessible_selectors"
   gem "cuprite", require: "capybara/cuprite"
+  gem "capybara_accessible_selectors", github: "opf/capybara_accessible_selectors", branch: "feature/cuprite-support"
```

If we're satisfied with introducing a load-order requirement, we should document it thoroughly (maybe in the `README.md`, maybe elsewhere).

myabc commented on spec/spec_helper.rb:
```diff
@@ -8,6 +8,11 @@
 require "debug"
 require "capybara/rspec"
 require "selenium-webdriver"
+begin
+  require "capybara/cuprite"
+rescue LoadError
+  # Cuprite is optional
+end
```
I'm not sure if this is necessary - it might be ok to make Cuprite (and all Capybara drivers) test dependencies.

myabc commented on lib/capybara_accessible_selectors/node/accessible_description.rb:
```diff
@@ -27,4 +27,13 @@ def accessible_description
+      # The accname-1.2 description algorithm is driver-agnostic (it only uses
+      # the generic Capybara node API), so reuse it instead of Chrome's raw
+      # computed description, keeping results consistent with Selenium
+      Selenium::AccessibleDescription.resolve(self) || ""
```
N.B. we might consider renaming this module to `Capybara::AccessibleDescription`

myabc committed:
Fix disclosure toggle on Cuprite driver

details[:open] is a boolean attribute whose raw value differs per driver
(Selenium returns "true"/nil, Cuprite returns true/false). The summary
toggle compared it against the desired state as a string, so on Cuprite a
re-toggle of an already-open disclosure misfired and closed it.

Normalise the value before comparing, and un-skip the two disclosure
specs that exercised re-toggling.

myabc committed:
Fix rich text clear on Cuprite driver

fill_in_rich_text cleared existing content with a select-all chord
(mod+a) followed by backspace. Cuprite/Ferrum cannot reliably trigger a
select-all chord inside a contenteditable, so the clear was a no-op and
new text was appended instead of replacing.

For Cuprite, move the caret to the end and backspace through the content
instead (matching the approach used elsewhere for Ferrum); keep the
chord for Selenium. Un-skip the eight rich text specs that exercised
clearing or replacing content.

myabc commented:
@seanpdoyle please give this a look when you get a chance! With the help of Claude I was able to improve robustness and work through the remaining skipped specs. All specs should now pass on Cuprite.

cnorthwood commented:
hi @myabc - just wanted to acknowledge this PR and let you know we're looking at this internally as we've not really had much in the way of outside contributors to this project, and the original maintainer of it is no longer working at Citizens Advice so we're going to find an internal developer to take over stewardship of this who can review the PR and get back to you. sorry you've been left in limbo a little bit!

myabc commented:
@cnorthwood no problem at all! thanks again for the great orject. Let me know if me and colleagues at @opf / OpenProject can do anything to support the project!

myabc commented on lib/capybara_accessible_selectors/selectors/rich_text.rb:
```diff
@@ -73,8 +71,8 @@ def fill_in_iframe_rich_text(frame, text, clear)
-        editable.send_keys [command_modifier, "a"], :backspace if editable.text != "" && clear
-        editable.send_keys text
+        clear_rich_text(editable) if editable.text != "" && clear
+        editable.send_keys text if text
```
these guards can be removed when https://github.com/rubycdp/cuprite/pull/313 is merged.
